### PR TITLE
:sparkles: Save new attributes to organisation admin Brevo contact [NGC-2860]

### PR DESCRIPTION
**Pull request by bjlaa**

### CODE REVIEW

**Review by johangirod — COMMENTED** [review]
Review with AI assistance (j'ai fait un peu de ping-pong)



## 🔴 Critical

### 1. Falsy value check filters out legitimate `0` values

**File**: `apps/server/src/adapters/brevo/client.ts` — lines ~687, ~699, ~704

```javascript
...(lastPollParticipantsCount
  ? { [Attributes.LAST_POLL_PARTICIPANTS_NUMBER]: lastPollParticipantsCount }
  : {}),
```

`0` is falsy in JavaScript, so an organisation with 0 polls, 0 completed simulations, or 0 participants will **never** have these attributes sent to Brevo. Brevo segmentation rules targeting `= 0` will never match.

**Fix** — use explicit null/undefined checks on all 4 conditional attributes:

```javascript
...(lastPollParticipantsCount != null
  ? { [Attributes.LAST_POLL_PARTICIPANTS_NUMBER]: lastPollParticipantsCount }
  : {}),
...(pollsCreatedCount != null
  ? { [Attributes.NUMBER_ORGANISATION_CREATED_POLLS]: pollsCreatedCount }
  : {}),
...(organisationSimulationsCompletedCount != null
  ? { [Attributes.NUMBER_ORGANISATION_COMPLETED_SIMULATIONS]: organisationSimulationsCompletedCount }
  : {}),
```

---

### 2. Test expectations out of sync with handler behavior after last commit

**Files**: All 6 modified spec files (`create-organisation.spec.ts`, `create-organisation-poll.spec.ts`, `delete-organisation-poll.spec.ts`, `update-organisation-poll.spec.ts`, `update-organisation.spec.ts`, `create-poll-simulation.spec.ts`)

The last commit `♻️ additional attributes are only in job` removed `lastPollParticipantsCount`, `pollsCreatedCount`, `organisationSimulationsCompletedCount`, and `organisationLastSimulationDate` from the event handler — but the tests still expect them in the Brevo API payload:

```javascript
// Example from create-organisation-poll.spec.ts — these keys are no longer sent by the handler:
LAST_POLL_PARTICIPANTS_NUMBER: 0,              // ❌
NUMBER_ORGANISATION_CREATED_POLLS: 1,           // ❌
NUMBER_ORGANISATION_COMPLETED_SIMULATIONS: 0,   // ❌
LAST_ORGANISATION_SIMULATION_DATE: undefined,   // ❌
```

Since the mock uses `expect(...).toEqual(expectBody)` (exact deep equality), these tests should be failing.

**Fix** — remove all 4 attributes from every `expectBody` block across the 6 test files. They are now only sent by the cron job, not the event handlers.

---

## 🟠 Major

### 3. Duplicated query filter in `getOrganisationSimulationInfo`

**File**: `apps/server/src/features/organisations/organisations.repository.ts` — lines ~831-856

The same WHERE clause is copy-pasted across `count` and `findFirst`:

```javascript
const [simulationCount, simulationPoll] = await Promise.all([
  session.simulationPoll.count({
    where: {
      simulation: { progression: 1 },
      pollId: { in: organisationPollIds },
    },
  }),
  session.simulationPoll.findFirst({
    where: {
      simulation: { progression: 1 },        // duplicated
      pollId: { in: organisationPollIds },    // duplicated
    },
    ...
  }),
])
```

If the filter logic changes, someone will likely update one and forget the other.

**Fix** — extract the common filter:

```javascript
const completedSimFilter = {
  simulation: { progression: 1 },
  pollId: { in: organisationPollIds },
}

const [simulationCount, simulationPoll] = await Promise.all([
  session.simulationPoll.count({ where: completedSimFilter }),
  session.simulationPoll.findFirst({
    where: completedSimFilter,
    orderBy: { createdAt: 'desc' },
    select: { createdAt: true },
  }),
])
```

---

### 4. Per-organisation queries + sequential processing will be slow at scale

**Files**: `apps/server/src/jobs/sync-brevo-contacts.ts` lines ~44-98 and `apps/server/src/features/organisations/organisations.repository.ts` lines ~814-871

Two problems compound in the cron loop:

**a) Extra DB round-trip to fetch poll IDs.** Before querying completed simulations, `getOrganisationSimulationInfo` does a separate `poll.findMany` just to collect IDs:

```javascript
const pollIds = await session.poll.findMany({
  where: { organisationId },
  select: { id: true },
})
const organisationPollIds = pollIds.map(({ id }) => id)
```

**b) Sequential iteration.** The cron job processes every organisation one by one — for each one: 3 DB queries + 1 Brevo API call. For 10 000 organisations that's ~40 000 sequential I/O operations.

```javascript
for await (const organisation of batchOrganisations) {
  // 3 DB queries → Brevo API call → next org...
}
```

**Fix** — a single raw SQL query can compute all 3 stats (`lastPollParticipantsCount`, `pollsCreatedCount`, `organisationSimulationsCompletedCount`, `organisationLastSimulationDate`) for **all** organisations in one round-trip. The codebase already hints at this approach in the existing comment at `organisations.repository.ts` lines ~721-731. Something like:

```sql
SELECT
  o.id as "organisationId",
  COUNT(DISTINCT p.id) as "pollsCreatedCount",
  COUNT(DISTINCT sp."simulationId") FILTER (WHERE s.progression = 1) as "simulationsCompletedCount",
  MAX(s.date) FILTER (WHERE s.progression = 1) as "lastSimulationDate",
  COALESCE((
    SELECT COUNT(*) FROM "SimulationPoll" WHERE "pollId" = (
      SELECT p2.id FROM "Poll" p2 WHERE p2."organisationId" = o.id ORDER BY p2."createdAt" DESC LIMIT 1
    )
  ), 0) as "lastPollParticipantsCount"
FROM "Organisation" o
LEFT JOIN "Poll" p ON p."organisationId" = o.id
LEFT JOIN "SimulationPoll" sp ON sp."pollId" = p.id
LEFT JOIN "Simulation" s ON s.id = sp."simulationId"
GROUP BY o.id
```

This collapses 3 × N DB queries into 1. Then the only per-organisation work left is the Brevo API call, which can be parallelized with a small concurrency pool (e.g. `p-limit` with concurrency of 10).

---

### 5. No tests for the `sync-brevo-contacts` cron job

**File**: `apps/server/src/jobs/sync-brevo-contacts.ts` (new file — 0 tests)

The cron job is the only code path that computes `pollsCreatedCount`, `organisationSimulationsCompletedCount`, and `organisationLastSimulationDate`. There are zero tests covering it.

**Fix** — extract the core logic into an exported, testable function:

```javascript
export const syncOrganisationToBrevo = async (organisation: ...) => { ... }
```

Then add integration tests covering at minimum:
- Org with no polls → `pollsCreatedCount: 0`, `organisationSimulationsCompletedCount: 0`, no date
- Org with polls but no completed simulations → counts > 0, simulations = 0
- Org with multiple polls and completed simulations → correct count + most recent date
- Org with no administrator → skipped gracefully
- Brevo API failure → error counted, continues to next org

---

### 6. No edge case tests for `getOrganisationSimulationInfo`

**File**: `apps/server/src/features/organisations/organisations.repository.ts` — lines ~814-871

The new function lacks dedicated unit tests. Untested scenarios include:
- Polls exist but 0 simulations are completed (`progression !== 1`)
- Soft-deleted simulations — are they filtered correctly?
- Multiple polls with interleaved simulations across them — is the most recent date truly across **all** polls?

**Fix** — add a unit test file (or extend existing tests) covering these edge cases.

---

### Summary

| # | Severity | What | Where |
|---|----------|------|-------|
| 1 | 🔴 Critical | `0` values silently filtered out | `brevo/client.ts` L687-715 |
| 2 | 🔴 Critical | Tests expect attributes handler no longer sends | 6 spec files |
| 3 | 🟠 Major | Duplicated WHERE clause | `organisations.repository.ts` L831-856 |
| 4 | 🟠 Major | Per-org queries + sequential loop = slow at scale | `sync-brevo-contacts.ts` L44-98 + `organisations.repository.ts` L814-871 |
| 5 | 🟠 Major | Cron job has zero tests | `sync-brevo-contacts.ts` |
| 6 | 🟠 Major | `getOrganisationSimulationInfo` missing edge case tests | `organisations.repository.ts` L814-871 |

**Comment by bjlaa** [comment]
Modifié pas mal de choses notamment la fonction qui récupère les stats avec du SQL (pas ultra à l'aise chaud que tu relises bien la query @johangirod ) et ajouté des tests

File: apps/server/src/jobs/__tests__/sync-brevo-contacts.spec.ts [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    organisationLastSimulationDate?: Date | null
+  } = {}
+) => ({
+  lastPollParticipantsCount: overrides.lastPollParticipantsCount ?? 0,
+  pollsCreatedCount: overrides.pollsCreatedCount ?? 0,
+  organisationSimulationsCompletedCount:
+    overrides.organisationSimulationsCompletedCount ?? 0,
+  organisationLastSimulationDate:
+    overrides.organisationLastSimulationDate ?? null,
+})
```
Comment (johangirod) [inline apps/server/src/jobs/__tests__/sync-brevo-contacts.spec.ts]: On est pas dans une logique E2E ici. Tu testes uniquement la partie « envoie à brevo ». C'est bien, mais c'est insuffisant, et ça ne correspond pas à la logique des tests du back. 

Il vaudrait mieux tester le scénario « sync organisation stat  » dans son entièreté.

File: apps/server/src/jobs/__tests__/sync-brevo-contacts.spec.ts [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        })
+      )
+
+      await syncOrganisationToBrevo(org, makeStats())
+
+      expect(logger.info).toHaveBeenCalledWith(
+        `Synced Brevo contact for organisation ${org.id} (${org.name})`
+      )
+    })
+  })
```
Comment (johangirod) [inline apps/server/src/jobs/__tests__/sync-brevo-contacts.spec.ts]: Ce cas d'usage doit continuer à être gérer via l'event handler `addOrUpdateBrevoContact`, pas par le job (qui synchronise uniquement les variables brevo)

File: apps/server/src/jobs/sync-brevo-contacts.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import { getOrganisationsBatchBrevoStats } from '../features/organisations/organisations.repository.js'
+import logger from '../logger.js'
+
+interface SyncResult {
+  processedCount: number
+  errorCount: number
+}
+
+export const syncOrganisationToBrevo = async (
+  organisation: {
```
Comment (johangirod) [inline apps/server/src/jobs/sync-brevo-contacts.ts]: Ce type doit déjà exister quelque part, non ?

File: apps/server/src/jobs/sync-brevo-contacts.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+export const runSync = async ({
+  session,
+}: {
+  session: Session
+}): Promise<SyncResult> => {
+  let processedCount = 0
+  let errorCount = 0
+
+  // Batch-load all organisation stats in a single query
```
Comment (johangirod) [inline apps/server/src/jobs/sync-brevo-contacts.ts]: On peut avoir une seule requête SQL pour tout ça

File: apps/server/src/jobs/sync-brevo-contacts.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    optedInForCommunications,
+    type: type as OrganisationType,
+    lastPollParticipantsCount: stats?.lastPollParticipantsCount,
+    pollsCreatedCount: stats?.pollsCreatedCount,
+    organisationSimulationsCompletedCount:
+      stats?.organisationSimulationsCompletedCount,
+    organisationLastSimulationDate:
+      stats?.organisationLastSimulationDate ?? undefined,
+  })
+
```
Comment (johangirod) [inline apps/server/src/jobs/sync-brevo-contacts.ts]: Plutôt que passer par une fonction qui fait des choses en plus, on restreint juste aux variables à modifier (soit celles calculées par getOrganisationsBatchBrevoStats) en appelant `addOrUpdateContact` https://github.com/incubateur-ademe/nosgestesclimat-site-nextjs/blob/d2369836eb60eb8d29c06d4681d044287088e620/apps/server/src/adapters/brevo/client.ts#L586

File: apps/server/src/jobs/sync-brevo-contacts.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        slug: true,
+        type: true,
+        administrators: {
+          select: {
+            user: {
+              select: {
+                id: true,
+                name: true,
+                email: true,
+                optedInForCommunications: true,
```
Comment (johangirod) [inline apps/server/src/jobs/sync-brevo-contacts.ts]: optedInForCommunications : Non necessaire

File: apps/server/src/jobs/sync-brevo-contacts.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      await syncOrganisationToBrevo(organisation, stats)
+      processedCount++
+    } catch (error) {
+      errorCount++
+      logger.error(
+        `Failed to sync Brevo contact for organisation ${organisation.id}`,
+        { error }
+      )
+    }
+  }
```
Comment (johangirod) [inline apps/server/src/jobs/sync-brevo-contacts.ts]: On ajoute une logique de retry ici (1 ou 2 retry)

**Review by johangirod — COMMENTED** [review]
C'est mieux ! J'ai encore quelques petits retours.

J'ai testé la requête SQL en prod, on est bon (~1min et 4500 lignes). 

Je peux reprendre d'ici si tu veux.

**Review by johangirod — APPROVED** [review]

**Review by johangirod — APPROVED** [review]
